lib/route/link/bridge.c

```diff
@@ -107,22 +117,110 @@ static inline struct bridge_data *bridge_data(struct rtnl_link *link)
 
 static void *bridge_alloc(struct rtnl_link *link)
 {
-	return calloc(1, sizeof(struct bridge_data));
+	struct bridge_data *bridge_data = calloc(1, sizeof(struct bridge_data));
+	nl_init_list_head(&bridge_data->mst_list);
+
+	return bridge_data;
 }
 
-static void *bridge_clone(struct rtnl_link *link, void *data)
+static struct mst_state_entry *mst_state_entry_alloc(void)
 {
-	struct bridge_data *bd;
+	struct mst_state_entry *entry;
+
+	entry = calloc(1, sizeof(*entry));
+	if (entry == NULL)
+		return NULL;
+
+	nl_init_list_head(&entry->list_node);
+
+	return entry;
+}
+
+static struct mst_state_entry *mst_state_entry_create(uint16_t msti,
+						      uint8_t state)
+{
+	struct mst_state_entry *entry;
+
+	entry = mst_state_entry_alloc();
+	if (entry == NULL)
+		return NULL;
 
-	if ((bd = bridge_alloc(link)))
-		memcpy(bd, data, sizeof(*bd));
+	entry->msti = msti;
+	entry->state = state;
 
-	return bd;
+	return entry;
+}
+
+static void mst_state_entry_del(struct mst_state_entry *entry)
+{
+	nl_list_del(&entry->list_node);
+	free(entry);
+}
+
+static void mst_list_clear(struct nl_list_head *mst_list)
+{
+	struct mst_state_entry *entry;
+	struct mst_state_entry *entry_safe;
+
+	nl_list_for_each_entry_safe(entry, entry_safe, mst_list, list_node)
+		mst_state_entry_del(entry);
 }
 
 static void bridge_free(struct rtnl_link *link, void *data)
 {
-	free(data);
+	struct bridge_data *bd = data;
+
+	mst_list_clear(&bd->mst_list);
+	free(bd);
+}
+
+static struct mst_state_entry *find_mst_state_entry(struct bridge_data *bd,
+						    uint16_t msti)
+{
+	struct mst_state_entry *entry;
+
+	nl_list_for_each_entry(entry, &bd->mst_list, list_node) {
+		if (entry->msti == msti)
+			return entry;
+	}
+
+	return NULL;
+}
+
+static struct mst_state_entry *
+mst_state_entry_clone(struct mst_state_entry *src)
+{
+	return mst_state_entry_create(src->msti, src->state);
+}
+
+static void *bridge_clone(struct rtnl_link *link, void *data)
+{
+	struct bridge_data *src_bd = (struct bridge_data *)data;
+	struct bridge_data *dst_bd;
+	struct mst_state_entry *entry;
+
+	dst_bd = calloc(1, sizeof(*dst_bd));
+	if (!dst_bd)
+		return NULL;
+
+	memcpy(dst_bd, src_bd, sizeof(*dst_bd));
+
+	nl_init_list_head(&dst_bd->mst_list);
+
+	nl_list_for_each_entry(entry, &src_bd->mst_list, list_node) {
+		struct mst_state_entry *entry_copy =
+			mst_state_entry_clone(entry);
+
+		if (!entry_copy) {
+			mst_list_clear(&dst_bd->mst_list);
+			free(dst_bd);
```

Comment:
Got it. I might have been hesitant because that required passing the link object as well (which I find odd anyway). The link that is passed is the partially cloned link which the bridge data will soon belong to. In practice the link parameter is ignored anyway but I'm unsure of the intended use for it.

Reply:
let's do
```diff
diff --git i/lib/route/link/bridge.c w/lib/route/link/bridge.c
index 94a29b1df6a7..0d7af5ba704a 100644
--- i/lib/route/link/bridge.c
+++ w/lib/route/link/bridge.c
@@ -170,12 +170,17 @@ static void mst_list_clear(struct nl_list_head *mst_list)
 		mst_state_entry_del(entry);
 }
 
+static void bridge_data_free(struct bridge_data *bd)
+{
+	mst_list_clear(&bd->mst_list);
+	free(bd);
+}
+
 static void bridge_free(struct rtnl_link *link, void *data)
 {
 	struct bridge_data *bd = data;
 
-	mst_list_clear(&bd->mst_list);
-	free(bd);
+	bridge_data_free(bd);
 }
 
 static struct mst_state_entry *find_mst_state_entry(struct bridge_data *bd,
@@ -216,7 +221,7 @@ static void *bridge_clone(struct rtnl_link *link, void *data)
 			mst_state_entry_clone(entry);
 
 		if (!entry_copy) {
-			bridge_free(link, dst_bd);
+			bridge_data_free(dst_bd);
 			return NULL;
 		}
 
```